python/singa/sonnx.py

```diff
@@ -0,0 +1,242 @@
+            a[str(i.output[0])] = autograd.softmax(a[str(i.input[0])])
+        elif (i.op_type == 'Add'):
+            if(str(i.input[1])[-1] == 'b'):
+                a[str(i.output[0])] = autograd.add_bias(a[str(i.input[0])], a[str(i.input[1])])
+            else:
+                a[str(i.output[0])] = autograd.add(a[str(i.input[0])],a[str(i.input[1])])
+        elif (i.op_type == 'MatMul'):
+            a[str(i.output[0])] = autograd.matmul(a[str(i.input[0])], a[str(i.input[1])])
+
+    loss = autograd.cross_entropy(a['Y'], target)
+    return loss
+
+
+
+def get_onnx_model(y, dy=None):
+    '''
+    from singa computational graph(loss) to get onnx model
+    Args:
+        y:loss of singa autograd
+    Return:
+        onnx model
+    '''
+    ######################
+    import onnx
+    from onnx import helper
+    from onnx import AttributeProto, TensorProto, GraphProto
+    from onnx import numpy_helper
+    import numpy as np
+    from singa import tensor
+    X = helper.make_tensor_value_info('X', TensorProto.FLOAT, [1, 2])
+    Y = helper.make_tensor_value_info('Y', TensorProto.FLOAT, [1, 2])
+    node = []
+    ######################
+
+    dependency = infer_dependency(y.creator)
+    #get singa computational graph dependency
+
+    assert y.size() == 1, 'y must be a Tensor with a single value;' \
+                          'size of y is % d' % y.size()
+
+    # by default the dy is a tensor with 1.0 for each sample;
+    if dy is None:
+        dy = float(1.0)
+    elif isinstance(dy, Tensor):
+        dy = dy.data
+    else:
+        dy = float(dy)
+
+    # ready is a queue of (operation, dy list)
+    ready = deque([(y.creator, (dy,))])
+    # using bfs iterative method to check all computational graph nodes
+    not_ready = {}  # mapping: op->[dy]
+    gradients = {}  # mapping: x->dx if x.stores_grad
+    if y.stores_grad:
+        gradients[y] = dy
+
+    while len(ready) > 0:
+        op, dys = ready.pop()
+        #check all children nodes of this node
+        if not op.requires_grad or isinstance(op, Dummy):
+            continue
+        # if not isinstance(op, tensor.Dummy):
+        dxs = op._do_backward(*dys)
+        ##############################
+        cur = str(op).split('.')[-1].split(' ')[0]
+        pre = str(op.src[0][0]).split('.')[-1].split(' ')[0]
+        cc = str(op).split(' ')[-1]
+        pc = str(op.src[0][0]).split(' ')[-1]
+        cstrd = str(op) + str(dependency[op])
+        pstrd = str(op.src[0][0]) + str(dependency[op.src[0][0]])
+        cstr = str(op)
+        pstr = str(op.src[0][0])
+        #get father node's name and children nodes' names
+
+        if (pre == 'Dummy'): pstr = pre = 'X'
+        # if it is father node is dummy, the node will be input x
+        if (op.param['name'] == 'LeakyRelu'):
+        # save autograd layers param name, and use here for saving onnx model
+            node = [onnx.helper.make_node('LeakyRelu', inputs=[pstr], outputs=[cstr], )] + node
```

Comment:
can create a dictionary to replace the if else check?
Otherwise, everytime a new Op is added, we have to change the code here.

Reply:
yes